### DISCUSSION TITLE
CI system improvements

### DISCUSSION THREAD

**Discussion by ethomson** [libgit2/libgit2#5881] [Ideas]
With https://github.com/libgit2/libgit2/pull/4655, and with some other build system work that I had looked at, I was reminded that the build system came up at the libgit2 summit and we had some ideas for improvement.

Right now we don't have _any_ regular builds on certain platforms, including SPARC and ARM.  SPARC is important because it's easy to introduce alignment issues that only affect those processors.  ARM is important because there are many projects using it.

I think that our goals should be:
1. Fast validation of pull requests
2. High confidence that pull requests build and tests pass on at least the "core" platforms
3. Regular build and test pass on all "supported" platforms

Basically, I'm proposing keeping our pull request validation pass as-is, defining "core" platforms as Windows, macOS, Linux.  Then doing "regular" builds on a larger set of platforms.  (Where "regular" is to be determined.)

Right now we have 15 builds that we kick off for every pull request.  I think that we can achieve the same level of confidence that we have now and reduce the number of variations that we build for every PR.  Proposed list of core platforms:

1. Windows x86 (with MSVC 2010)
2. Windows amd64 (with MSVC 2010)
3. Linux amd64 (with gcc)
4. Linux amd64 (with clang)
5. macOS amd64

Enable memory leak detection on some subset of those builds.  Build some subset of those builds non-threadsafe.  (I think that subset should be "all" and "none" respectively, but let's discuss that.)  Build some subset of the Linux with OpenSSL and some subset with mbedTLS.

Then on a regular (nightly?  after merge into master?) basis, run an exhaustive set of build/test:

1. Windows x86 (with MSVC 2017) (it's hard for me to imagine a build failure upgrading MSVC, but it's not impossible, so I guess we should include it)
2. Windows amd64 (with MSVC 2017) - same as above
3. Windows mingw64 (32 bit)
4. Windows mingw64 (64 bit)
5. iOS (can we run the tests in an emulator?  @palmin or @nerdishbynature, help!)
6. FreeBSD amd64
7. FreeBSD SPARC
8. Raspberry Pi

Questions:
1. Excluding for a moment what is "core" and what is not, does this sound like a reasonable plan to split into a two phase system?
2. If "yes" to question 1, what is core?
3. If "yes" to question 1, what else is not core?

**Comment by carlosmn** [comment]
I think it does make sense to separate the "must work to merge" vs "must work to release" as a balance between the reaction speed from CI and the large range of platforms that are out there.

It would be useful to be able to trigger non-core ones for PRs as well, particularly in the cases where we're dealing with platform-specific things (or if we want to mess with threading).

Your list for "core" makes sense. the question would be which version of those systems we'd want to run (especially as newer macOS installs will have APFS instead of HFS+). For Linux we can presumably download whatever container and run it.

Do  you have ideas on how to run the FreeBSD and rpi builds?

**Comment by tiennou** [comment]
Agreed on your "core" list with "all" leak detection, "none" non-threadsafe.

> 5. iOS (can we run the tests in an emulator?

I don't see why not. Objective-Git has its own API test suite, but I'm pretty sure it's doable using the simulator. It's just that distributing for iOS requires heavy-lifting from the build system (4 archs).

> especially as newer macOS installs will have APFS instead of HFS+

I don't think it's possible, because Travis is still on 10.12, unless we manually manage Apple builds.

# 

There's also the special case of Coverity, which I'd keep in the "extended" list (ie. nightly/master-merge), along with the fuzzer support.

Maybe we can do the following CI dispatch, as that would make platform-specific CI failures obvious :

- GitLab CI: anything that can be containerized
- Travis: macOS & iOS
- AppVeyor: Windows

**Comment by ethomson** [comment]
High on my list of priorities is a single build platform that is ideally low or no cost.  Right now, I'm investigating the VSTS build system, which we rejected until recently because you had to be logged in to see build reports.  They[1] have recently added support for public / anonymous access to build reports.  VSTS has build agents hosted in Azure for Windows, Mac and Linux.  It also supports on-premises build agents (if, eg, we wanted to do nightlies for FreeBSD, or on a Raspberry Pi).  And it's free.  I've got preliminary support for Windows (MSVC) / Mac / Linux now, which is quite simple, so now I'm looking at expanding the list to include mingw64 on Windows (I'm looking at both a container-based approach and the current download-and-unpack-mingw64 approach - although I prefer the container-based approach in theory, it involves more downloading).

[1]: Note that for the purposes of full disclosure, I'm a Microsoft employee.

**Comment by ethomson** [comment]
Anyway, I hope to have a preliminary look at this over the next week so that we can review it and decide if it's a system we want to continue pursuing.

**Comment by tiennou** [comment]
If VSTS allows CI to be kept in one place, I'm 👍, though I don't really mind having 3 separate CIs (as I proposed above, which makes it easy to tell at a glance what "platform" is failing, while also giving some redundancy against transient failures).

**Comment by pks-t** [comment]
Some random thoughts:

- I don't know whether mbedTLS should be part of our "core" set of supported variations. I might be mistaken, but I don't have the feeling that it's in frequent use except for some niche platforms
- I'd like to enable memory leak detection on all platforms. On Windows it's basically free, and if we reduce the build matrix for other systems we probably wouldn't be slowed down by that
- I think non-threadsafe builds shouldn't be part of our core set of supported variations, and thus as you said the subset that builds like that should be "none". All modern operating systems that we target should have primitves to support threading
- I'd like to add a system to the non-core set with libressl. I guess the most reasonable choice  would be OpenBSD there

I think the split into two phases is perfectly reasonable. We have to improvise at some point without loosing too much, so this is the best way forward, I guess. I don't think having that run as a nightly job makes sense, though, but rather run it for each merge to master/maint. Ideally, we'd like to know that each first-parent commit on these branches works for all core and non-core platforms.

**Comment by carlosmn** [comment]
You don't need multiple providers to get different checks. I don't know if VSTS lets you do this easily, but you can set as many statuses for a commit as you like. You can set one per build configuration without having to have different providers for each.

@tiennou I'm not sure how you got to your list, but AppVeyor and Travis for macOS are the first ones I'd like to get rid of as they're by far the slowest.

**Comment by tiennou** [comment]
> AppVeyor and Travis for macOS are the first ones I'd like to get rid of as they're by far the slowest.

I was just suggesting given what we already have, and kept those because they're the only ones which support Windows or macOS (as those 2 platforms can't be containerized). If VSTS allows us to get rid of them, and they're faster, I'm 👍. It's just that I liked being able to tell at a glance, given 3 platform-bound CIs, which ones are failing (but I think that's what you mean by "you can set as many statuses for a commit"). If so, I'm also 👍.

> I'd like to enable memory leak detection on all platforms.

We're only missing out on macOS, because the collection happens "out of band" (you set MALLOC_STACK_LOGGING_NO_COMPACT=1 and call `leaks $clarpid`). It just requires some changes to clar so it sticks around before exiting at the end of the run.
